File: 14_stacks_and_queues_2/08_stock_span.py
class StockSpanner:
    def __init__(self):
        self.prices = []
        self.pge = []

    def next(self, price: int) -> int:
        self.prices.append(price)
        n = len(self.prices)

        while self.pge and self.prices[self.pge[-1]] <= price:
            self.pge.pop()

        res = n
        if self.pge:
            res = n - self.pge[-1] - 1

        self.pge.append(n - 1)

        return res

File: 14_stacks_and_queues_2/08_stock_span.rs
struct StockSpanner {
    stocks: Vec<i32>,
    pge: Vec<i32>,
}

impl StockSpanner {
    fn new() -> Self {
        StockSpanner {
            stocks: Vec::new(),
            pge: Vec::new(),
        }
    }

    fn next(&mut self, price: i32) -> i32 {
        self.stocks.push(price);
        let n = self.stocks.len() as i32;

        while !self.pge.is_empty() && self.stocks[*self.pge.last().unwrap() as usize] <= price {
            self.pge.pop();
        }
        let mut res = n;

        if !self.pge.is_empty() {
            res = n - *self.pge.last().unwrap() - 1;
        }
        self.pge.push(n - 1);

        res
    }
}
